File: ultrasonic/src/main/java/org/moire/ultrasonic/util/AlbumHeader.java
package org.moire.ultrasonic.util;

import android.content.Context;

import org.moire.ultrasonic.domain.MusicDirectory;

import java.util.HashSet;
import java.util.Set;

public class AlbumHeader
{
	private boolean isAllVideo;
	private long totalDuration;
	private Set<String> artists;
	private Set<String> grandParents;
	private Set<String> genres;
	private Set<Integer> years;

	public boolean getIsAllVideo()
	{
		return isAllVideo;
	}

	public long getTotalDuration()
	{
		return totalDuration;
	}

	public Set<String> getArtists()
	{
		return artists;
	}

	public Set<String> getGrandParents()
	{
		return this.grandParents;
	}

	public Set<String> getGenres()
	{
		return this.genres;
	}

	public Set<Integer> getYears()
	{
		return this.years;
	}

	public AlbumHeader()
	{
		this.artists = new HashSet<String>();
		this.grandParents = new HashSet<String>();
		this.genres = new HashSet<String>();
		this.years = new HashSet<Integer>();

		this.isAllVideo = true;
		this.totalDuration = 0;
	}

	public static AlbumHeader processEntries(Context context, Iterable<MusicDirectory.Entry> entries)
	{
		AlbumHeader albumHeader = new AlbumHeader();

		for (MusicDirectory.Entry entry : entries)
		{
			if (!entry.isVideo())
			{
				albumHeader.isAllVideo = false;
			}

			if (!entry.isDirectory())
			{
				if (Util.shouldUseFolderForArtistName())
				{
					albumHeader.processGrandParents(entry);
				}

				if (entry.getArtist() != null)
				{
					Integer duration = entry.getDuration();

					if (duration != null)
					{
						albumHeader.totalDuration += duration;
					}

					albumHeader.artists.add(entry.getArtist());
				}

				if (entry.getGenre() != null)
				{
					albumHeader.genres.add(entry.getGenre());
				}

				if (entry.getYear() != null)
				{
					albumHeader.years.add(entry.getYear());
				}
			}
		}

		return albumHeader;
	}

	private void processGrandParents(MusicDirectory.Entry entry)
	{
		String grandParent = Util.getGrandparent(entry.getPath());

		if (grandParent != null)
		{
			this.grandParents.add(grandParent);
		}
	}
}

File: ultrasonic/src/main/kotlin/org/moire/ultrasonic/adapters/AlbumHeader.kt
package org.moire.ultrasonic.adapters

import java.util.HashSet
import org.moire.ultrasonic.domain.Identifiable
import org.moire.ultrasonic.domain.MusicDirectory
import org.moire.ultrasonic.util.Settings.shouldUseFolderForArtistName
import org.moire.ultrasonic.util.Util.getGrandparent

class AlbumHeader(
    var entries: List<MusicDirectory.Child>,
    var name: String?
) : Identifiable {
    var isAllVideo: Boolean
        private set

    var totalDuration: Long
        private set

    var childCount = 0

    private val _artists: MutableSet<String>
    private val _grandParents: MutableSet<String>
    private val _genres: MutableSet<String>
    private val _years: MutableSet<Int>

    val artists: Set<String>
        get() = _artists

    val grandParents: Set<String>
        get() = _grandParents

    val genres: Set<String>
        get() = _genres

    val years: Set<Int>
        get() = _years

    private fun processGrandParents(entry: MusicDirectory.Child) {
        val grandParent = getGrandparent(entry.path)
        if (grandParent != null) {
            _grandParents.add(grandParent)
        }
    }

    @Suppress("NestedBlockDepth")
    private fun processEntries(list: List<MusicDirectory.Child>) {
        entries = list
        childCount = entries.size
        for (entry in entries) {
            if (!entry.isVideo) {
                isAllVideo = false
            }
            if (!entry.isDirectory) {
                if (shouldUseFolderForArtistName) {
                    processGrandParents(entry)
                }
                if (entry.artist != null) {
                    val duration = entry.duration
                    if (duration != null) {
                        totalDuration += duration.toLong()
                    }
                    _artists.add(entry.artist!!)
                }
                if (entry.genre != null) {
                    _genres.add(entry.genre!!)
                }
                if (entry.year != null) {
                    _years.add(entry.year!!)
                }
            }
        }
    }

    init {
        _artists = HashSet()
        _grandParents = HashSet()
        _genres = HashSet()
        _years = HashSet()

        isAllVideo = true
        totalDuration = 0

        processEntries(entries)
    }

    override val id: String
        get() = "HEADER"

    override val longId: Long
        get() = -1L

    override fun compareTo(other: Identifiable): Int {
        return this.longId.compareTo(other.longId)
    }
}
